lib/src/command/outdated.dart

```diff
@@ -0,0 +1,440 @@
+      rows.retainWhere(
+          (r) => (r.current ?? r.upgradable)?.version != r.latest?.version);
+    }
+    if (!includeDevDependencies) {
+      rows.removeWhere((r) => r.kind == _DependencyKind.dev);
+    }
+
+    rows.sort();
+
+    if (argResults['format'] == 'json') {
+      // TODO(sigurdm): decide on and document the json format.
+      await _outputJson(rows);
+    } else {
+      final useColors = argResults['format'] == 'color' ||
+          (!argResults.wasParsed('format') && stdin.hasTerminal);
+      final marker = {
+        'outdated': _OutdatedMarker(),
+        'none': _NoneMarker(),
+      }[argResults['mark']];
+      await _outputHuman(rows, marker, useColors);
+    }
+  }
+
+  /// Retrieves the pubspec of package [name] in [version] from [source].
+  Future<Pubspec> _describeVersion(
+      String name, Source source, dynamic description, Version version) async {
+    return version == null
+        ? null
+        : await cache
+            .source(source)
+            .describe(PackageId(name, source, version, description));
+  }
+}
+
+Pubspec _stripDevDependencies(Pubspec original) {
+  return Pubspec(
+    original.name,
+    sdkConstraints: original.sdkConstraints,
+    dependencies: original.dependencies.values,
+    devDependencies: null,
+  );
+}
+
+/// Returns new pubspec with the same dependencies as [original] but with no
+/// version constraints on hosted packages.
+Pubspec _stripVersionConstraints(Pubspec original) {
+  List<PackageRange> _unconstrained(Map<String, PackageRange> constrained) {
+    final result = <PackageRange>[];
+    for (final name in constrained.keys) {
+      final packageRange = constrained[name];
+      var unconstrainedRange = packageRange;
+      if (packageRange.source is HostedSource) {
+        unconstrainedRange = PackageRange(
+            packageRange.name,
+            packageRange.source,
+            VersionConstraint.any,
+            packageRange.description,
+            features: packageRange.features);
+      }
+      result.add(unconstrainedRange);
+    }
+    return result;
+  }
+
+  return Pubspec(
+    original.name,
+    sdkConstraints: original.sdkConstraints,
+    dependencies: _unconstrained(original.dependencies),
+    devDependencies: _unconstrained(original.devDependencies),
+    // TODO(sigurdm): consider dependency overrides.
+  );
+}
+
+Future<void> _outputJson(List<_PackageDetails> rows) async {
+  log.message(JsonEncoder.withIndent('  ')
+      .convert({'packages': rows.map((row) => row.toJson()).toList()}));
+}
+
+Future<void> _outputHuman(
+    List<_PackageDetails> rows, _Marker marker, bool useColors) async {
+  if (rows.isEmpty) {
+    log.message('Found no outdated packages');
+    return;
+  }
+
+  final formattedRows = <FormattedRow>[
+    FormattedRow(
+        null,
+        ['Package', 'Current', 'Upgradable', 'Resolvable', 'Latest']
+            .map((s) => _FormattedString(s, format: log.bold))
+            .toList()),
```

Comment:
I would suggest adding:
```dart
...rows.where((r) => r.kind != _DependencyKind.production).map(marker.formatRow)
FormattedRow(['dev_dependencies', ]),
...rows.where((r) => r.kind != _DependencyKind.dev).map(marker.formatRow)
```

Or something like this... The ` if (lastKind != row.kind) {` is a bit sketchy :)

Reply:
You can still do this an async manner..